net/gnrc/rpl: fix default route lifetime

haukepetersen opened the pull request:
### Contribution description
Running some IP-over-BLE experiments using RPL, I stumbled over some packet loss which at first I could not explain. On random basis, every now an then a node started dropping IP packets for a short amount of time (~4-5s), although there was no packet loss on the link layer. 

After some digging, I found this was caused by be default route being removed for a short amount of time. If a node tried to send an IP packet to its parent node exactly in that time frame, where the default route entry was gone, this packet was dropped. However, since there was no packet loss on the link layer, the default route should never be removed in the first place....

After some more digging I was able to find the reason for this: when a node receives a DIO message from its parent, it sets a PARENT_TIMEOUT timer and updates the forwarding table (FT) entry of the default route with a given lifetime. The problem at hand is, that the current implementation uses exactly the same value (default is 300s) for both the PARENT_TIMEOUT timer AND the FT entry lifetime. Once this time runs up, the default route is removed from the FT and the PARENT_TIMEOUT event are is triggered at the same time. But to refresh the default route, RPL needs to probe the parent (by sending a DIS) and await and parse the response (DIO). 

So while waiting for the DIO to arrive, the default route is not set, leading to the faulty behavior described on top. Keeping in mind, that DIS or DIO messages might as well be lost on lossy links, the time span without a valid default route is even prolonged, increasing the potential packet loss...

The solution is quite simple: when setting the default route's lifetime, we simply need to add the time it takes the RPL node to probe its parent to the default lifetime. In particular, the probing may take a maximum of `GNRC_RPL_PARENT_TIMEOUT * GNRC_RPL_PARENT_PROBE_INTERVAL (-> 5 * 2000)`us. So this PR adds that time on top of the default lifetime of `CONFIG_GNRC_RPL_DEFAULT_LIFETIME * CONFIG_GNRC_RPL_DEFAULT_LIFETIME (5 * 60)`s.

Bonus: this PR also fixes a time scale conversion bug in `gnrc_rpl_parent_remove()`: `gnrc_ipv6_nib_ft_add()` takes the lifetime in seconds, but in `gnrc_rpl_parent_remove()` it was given in `us`. Seems this code was never tested :-)


### Testing procedure
Probably build test + release tests for RPL.

On top, here are the before and after packet delivery rates from my IP-over-BLE experiment:
Master:
[em7_uni_statconn-rpl_1s1h59b_20200824-100524_pdr.pdf](https://github.com/RIOT-OS/RIOT/files/5118588/em7_uni_statconn-rpl_1s1h59b_20200824-100524_pdr.pdf)


With this fix:
[em7_uni_statconn-rpl_1s1h59b_20200824-143116_pdr.pdf](https://github.com/RIOT-OS/RIOT/files/5118586/em7_uni_statconn-rpl_1s1h59b_20200824-143116_pdr.pdf)

### Issues/PRs references
none

benpicco commented on sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c:
```diff
@@ -66,6 +66,15 @@ static void _rpl_trickle_send_dio(void *args)
 
+/* The lifetime of the default route should exceed the parent timeout interval
+ * by the time we allow the node to probe its parent */
+static uint16_t _dflt_route_lifetime(gnrc_rpl_dodag_t *dodag)
```
What unit is the return type of this function?
maybe use `_dflt_route_lifetime_ms()`

haukepetersen replied:
can do.

benpicco commented on sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c:
```diff
@@ -66,6 +66,15 @@ static void _rpl_trickle_send_dio(void *args)
+static uint16_t _dflt_route_lifetime(gnrc_rpl_dodag_t *dodag)
+{
+    return (dodag->default_lifetime * dodag->lifetime_unit) +
+           (GNRC_RPL_PARENT_TIMEOUT *
+            (GNRC_RPL_PARENT_PROBE_INTERVAL / MS_PER_SEC));
```
If the function should return the time in ms, why divide by `MS_PER_SEC` here?

haukepetersen replied:
It should not, it returns `s` but `GNRC_RPL_PARENT_PROBE_INTERVAL` is defined in `ms` :-)

benpicco replied:
That's why I prefer to have the unit encoded in the name of the define :)

haukepetersen commented:
pushed a fix to append `_sec` to the newly introduced static function

benpicco commented:
Please squash!
How did you create the test setup btw?

haukepetersen commented:
squashed.

My setup consists of 15 nrf52dk and nrf52840dk nodes in the iotlab, runnig IP-over-BLE using NimBLE and a selection of connection manager implemenations (e.g. `nimble_autoconn`, `nimble_statconn`). On top some RPL-over-BLE (work in progress: https://github.com/haukepetersen/RIOT/tree/add_nimble_rpble).

So far, all my test setup + analysis tooling is not yet public, but I will publish it as soon as I have some progress on my results and once I cleaned it up a little :-)

benpicco reviewed (approved):
Reasoning is sound and the code looks good.

miri64 commented:
And go!
